feat: provide repository contributors insights

cecobask opened the pull request:
## Description
This pull request adds the feature to provide repository contributors insights.


## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents
Closes #5

## Mobile & Desktop Screenshots/Recordings

### File contents:
```
$ cat repos.yaml

- https://github.com/kubernetes/kubernetes
- https://github.com/cli/cli
```

### Usage with `--file` / `-f` flag
```
$ pizza insights contributors --file repos.yaml

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Repository URL      ┃ https://github.com/cli/cli ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ New contributors    ┃ 25                         ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Recent contributors ┃ 30                         ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Churn contributors  ┃ 40                         ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Repeat contributors ┃ 5                          ┃
┗━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Repository URL      ┃ https://github.com/kubernetes/kubernetes ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ New contributors    ┃ 158                                      ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Recent contributors ┃ 280                                      ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Churn contributors  ┃ 197                                      ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Repeat contributors ┃ 122                                      ┃
┗━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
```

### Usage with `--repos` / `-r` flag
```
$ pizza insights contributors --repos "https://github.com/kubernetes/kubernetes,https://github.com/cli/cli"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Repository URL      ┃ https://github.com/cli/cli ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ New contributors    ┃ 25                         ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Recent contributors ┃ 30                         ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Churn contributors  ┃ 40                         ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Repeat contributors ┃ 5                          ┃
┗━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Repository URL      ┃ https://github.com/kubernetes/kubernetes ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ New contributors    ┃ 158                                      ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Recent contributors ┃ 280                                      ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Churn contributors  ┃ 197                                      ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Repeat contributors ┃ 122                                      ┃
┗━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
```

### Ignore non-existent and non-indexed repositories
```
$ pizza insights contributors --repos "https://github.com/fake-owner/fake-repo,https://github.com/kubernetes/kubernetes"

ignoring repository issue: repository https://github.com/fake-owner/fake-repo is either non-existent or has not been indexed yet
┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Repository URL      ┃ https://github.com/kubernetes/kubernetes ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ New contributors    ┃ 158                                      ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Recent contributors ┃ 280                                      ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Churn contributors  ┃ 197                                      ┃
┣━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ Repeat contributors ┃ 122                                      ┃
┗━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
```

### Error when neither `--file` or `--repos` flag was provided
```
$ pizza insights contributors

Error: must pass either --file or --repos flag
Usage:
  pizza insights contributors [flags]

Flags:
  -f, --file string     Path to yaml file containing an array of git repository urls
  -h, --help            help for contributors
  -p, --period int      Number of days, used for query filtering (default 30)
  -r, --repos strings   Comma-separated array of git repository urls

Global Flags:
      --beta              Shorthand for using the beta OpenSauced API endpoint ("https://beta.api.opensauced.pizza/v1"). supercedes the '--endpoint' flag
  -e, --endpoint string   The API endpoint to send requests to (default "https://api.opensauced.pizza/v1")

2023/08/20 19:36:16 must pass either --file or --repos flag
```

### Error when both `--file` and `--repos` flags were provided
```
$ pizza insights contributors --file repos.yaml --repos "https://github.com/kubernetes/kubernetes,https://github.com/cli/cli"

Error: if any flags in the group [file repos] are set none of the others can be; [file repos] were all set
Usage:
  pizza insights contributors [flags]

Flags:
  -f, --file string     Path to yaml file containing an array of git repository urls
  -h, --help            help for contributors
  -p, --period int      Number of days, used for query filtering (default 30)
  -r, --repos strings   Comma-separated array of git repository urls

Global Flags:
      --beta              Shorthand for using the beta OpenSauced API endpoint ("https://beta.api.opensauced.pizza/v1"). supercedes the '--endpoint' flag
  -e, --endpoint string   The API endpoint to send requests to (default "https://api.opensauced.pizza/v1")

2023/08/20 19:36:42 if any flags in the group [file repos] are set none of the others can be; [file repos] were all set
```

## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [x] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [ ] 🙅 no documentation needed

jpmcb reviewed (changes requested):
I'm realizing this has quite abit of overlap with #24.

We'll want to align the user experience so that there is a clear story for CLI users with each command

This will also be blocked on us getting an API client for Go: https://github.com/open-sauced/pizza-cli/issues/28

brandonroberts commented on pkg/api/constants.go:
```diff
@@ -1,6 +1,7 @@
 const (
-	APIEndpoint     = "https://api.opensauced.pizza/v1"
-	BetaAPIEndpoint = "https://beta.api.opensauced.pizza/v1"
+	AlphaAPIEndpoint = "https://alpha.api.opensauced.pizza/v1"
```
Not often. We used it before when prototyping against different infrastructure, but haven't needed to use it since then. I don't think we need to include it

k1nho commented on cmd/insights/contributors.go:
```diff
@@ -0,0 +1,309 @@
+			repoContributorInsights.Repeat = append(repoContributorInsights.Repeat, data.AuthorLogin)
+		}
+	}()
+	go func() {
```
Very cool concurrency going on here! have been looking into also doing this on #24 as well to parallelize the data collection of the contributors so that the TUI can get the initial model started faster. Some questions if you don't mind. For this part, could it be the case that this is a dead goroutine when an error is returned from one of the data collectors, like the select is blocking once the error is acquired from ```errorChan``` closes the channel and then returns nil and err but waitGroup is not done, so ```doneChan``` does not get close?

Also,  could this be a good case to use [errGroup](https://pkg.go.dev/golang.org/x/sync@v0.3.0/errgroup)? @jpmcb

cecobask replied:
Hi @k1nho, thanks!
The `waitGroup` will always close, regardless if an error occurred or not. This is because the start of each go routine explicitly defers the call to `waitGroup.Done()`.
In my opinion, this is a suitable place to use `errGroup`. It just comes down to style/preference. Behind the hood, they're doing something similar if I'm not wrong.

bdougie commented:
> I'm realizing this has quite abit of overlap with #24.
> 
> We'll want to align the user experience so that there is a clear story for CLI users with each command

I see the difference being the interactivity you get from the show. I didn't want to increase scope on #24 while we explore the basic interaction, but think we can do some more high level insights from commits and prs as well.

jpmcb commented:
@cecobask - the Go API client has landed ✨ https://github.com/open-sauced/go-api !!!! Are you able to give that a try? It likely has some sharp edges so please open issues in that repo as the need arises!

jpmcb commented:
Can you change this PR's target branch to be `beta`? 👀 We recently changed the default branch to `beta` in support of using the releaser action

For further reading:
- https://github.com/open-sauced/pizza-cli/issues/34

cecobask commented:
@jpmcb, Done 👍🏻 Another review session is in order 😄 
- Refactored the codebase to make API calls using the go client where possible, the only exception being `auth` and `repo-query` commands
- Consolidated duplicated logic and constants
- Configured `gci` linter with default settings
- Added concurrency to the `pizza bake` command
- Updated `pizza insights contributors` to use https://github.com/charmbracelet/bubbles#table for outputting results

Until https://github.com/open-sauced/go-api/issues/2 is resolved, `pizza bake` commands will fail with `400 Bad Request`.

cecobask commented:
The problems with [open-sauced/go-api](https://github.com/open-sauced/go-api/issues/2) were resolved.
I believe there are no more changes to be added here.
@jpmcb, can you please take a look when you get a chance?

cecobask commented on main.go:
```diff
@@ -12,11 +13,9 @@ func main() {
-
 	err = rootCmd.Execute()
 	if err != nil {
-		log.Fatal(err)
```
Removed this log entry since the `Execute` method already outputs the error.
Previously, the error was outputted twice.

cecobask commented on cmd/root/root.go:
```diff
@@ -3,31 +3,50 @@ package root
+			}
+			return nil
+		},
+		SilenceUsage: true,
```
Every time a command returned an error, the `help` output was also printed, together with the error.
I think this is cleaner. The help output will still be shown in case the user entered invalid input for commands or passed `--help` flag.

jpmcb replied:
> Every time a command returned an error, the help output was also printed

We should still dump the help message when an error is returned since it's usually due to some misconfiguration of flags or values. The standard approach in Cobra CLIs is to print the help message along with the error so the user can have a quick loop to correcting it.

jpmcb commented on cmd/repo-query/repo-query.go:
```diff
@@ -49,29 +46,23 @@ func NewRepoQueryCommand() *cobra.Command {
 			return run(opts)
 		},
 	}
-
 	cmd.Flags().StringVarP(&opts.branch, "branch", "b", "HEAD", "The branch to index")
-
```
nit: Spacing was for readability

jpmcb commented on cmd/repo-query/repo-query.go:
```diff
@@ -85,7 +76,7 @@ func newRepoQueryAgent(apiClient *api.Client) *repoQueryAgent {
 }
 
-func (rq *repoQueryAgent) getOwnerAndRepo(url string) (owner, repo string, err error) {
+func GetOwnerAndRepoFromURL(url string) (owner, repo string, err error) {
```
Looks like it's getting used in `contributors.go` which is why it was changed from a method on the `repoQueryAgent` struct to a publicly facing package function.

This is a pretty general utility function, so we should abstract it out into some a utility package in `pkg/`. Maybe something like `pkg/github` since it's pretty specific to GitHub

jpmcb commented on cmd/root/root.go:
```diff
@@ -4,30 +4,49 @@ package root
+	"github.com/spf13/cobra"
 )
 
+const ()
+
```
What's this empty constant block for?

jpmcb commented on pkg/api/constants.go:
```diff
@@ -1,6 +1,46 @@
+	BetaEndpointKey
+	AlphaEndpointKey
+	InvalidEndpointKey   = 666
+	LocalhostEndpointURL = "http://localhost:3001"
+	ProdEndpointURL      = "https://api.opensauced.pizza"
+	BetaEndpointURL      = "https://beta.api.opensauced.pizza"
+	AlphaEndpointURL     = "https://alpha.api.opensauced.pizza"
+	ToolsEndpointURL     = "https://opensauced.tools"
```
Thanks for adding these!! 👍🏼

jpmcb commented on pkg/api/constants.go:
```diff
@@ -1,6 +1,46 @@
+	ToolsEndpointURL     = "https://opensauced.tools"
 )
+
+func EndpointKeyToURL(endpointKey int) (string, error) {
+	switch endpointKey {
+	case LocalhostEndpointKey:
+		return LocalhostEndpointURL, nil
+	case ProdEndpointKey:
+		return ProdEndpointURL, nil
+	case BetaEndpointKey:
+		return BetaEndpointURL, nil
+	case AlphaEndpointKey:
+		return AlphaEndpointURL, nil
+	default:
+		return "", fmt.Errorf("unknown endpoint key %d", endpointKey)
+	}
+}
```
I'm generally not a fan of this approach. It always seems redundant. Why not just have callers get the endpoints directly through imports?

```go
myEndpoint := api.BetaEndpointURL
```

and I'm abit confused why we would need this? Generally, if a user provides `--endpoint`, we should take that whole sale and not check it against our keying of endpoints hard coded into the CLI. I may deploy my own API behind a local network and I want to provide that explicitly like:

```
pizza --endpoint https://api.opensauced.my-local-network.homelab ...
```

cecobask replied:
For some reason, I assumed the go api client can only be used with the server configurations that it originally has.
I've refactored it to allow the use of any endpoint url.

jpmcb commented on pkg/utils/root.go:
```diff
@@ -2,11 +2,10 @@ package utils
 // user experience pieces to the root cobra command
 func SetupRootCommand(rootCmd *cobra.Command) {
 	cobra.AddTemplateFunc("wrappedFlagUsages", wrappedFlagUsages)
-
```
nit: This spacing was for readability

jpmcb commented on pkg/api/client.go:
```diff
@@ -18,3 +24,15 @@ func NewClient(endpoint string) *Client {
+	configuration := client.NewConfiguration()
+	apiClient := client.NewAPIClient(configuration)
+	context := context.WithValue(context.Background(), client.ContextServerIndex, endpointKey)
+	return context, apiClient, nil
```
Why return the context? The background context should be handled by the API if for some reason it had a panic or exited early. And our control flow shouldn't be handling the background context anyways.

I wonder if it'd be better to do `context.TODO` since we're not really handling different context's control flow right now.

jpmcb commented on cmd/insights/contributors.go:
```diff
@@ -0,0 +1,289 @@
+	// Repos is the array of git repository urls
+	Repos []string
+	// File is the path to yaml file containing an array of git repository urls
+	File string
```
nit: `FilePath` field makes more sense to me.

jpmcb commented on cmd/insights/contributors.go:
```diff
@@ -0,0 +1,289 @@
+	cmd.Flags().StringVarP(&opts.File, common.FlagNameFile, "f", "", "Path to yaml file containing an array of git repository urls")
+	cmd.Flags().StringSliceVarP(&opts.Repos, common.FlagNameRepos, "r", nil, "Comma-separated array of git repository urls")
+	cmd.Flags().Int32VarP(&opts.Period, common.FlagNamePeriod, "p", 30, "Number of days, used for query filtering")
+	cmd.MarkFlagsMutuallyExclusive(common.FlagNameFile, common.FlagNameRepos)
```
Same as in the bake command, I believe these flags being mutually exclusive is confusing. Instead, I would like to see:

```
pizza insights contributors URL
```
```
pizza insights contributors URL --file my-other-repos.yaml
```
```
pizza contributors insights --file my-repos.yaml
```

jpmcb commented on cmd/insights/contributors.go:
```diff
@@ -0,0 +1,289 @@
+			repoContributorInsights.RenderTable()
+		}(repoURL)
+	}
+	go func() {
+		waitGroup.Wait()
+		close(doneChan)
+	}()
```
This looks like a similar structuring as the bake command refactor you have: I don't think having the `waitgroup.Wait()` in a separate thread make sense. The main through should just wait for these all to finish.

And similar comments to the `errorChan`: these will get squashed and only the first error will surface and be returned.

jpmcb commented on cmd/insights/contributors.go:
```diff
@@ -0,0 +1,289 @@
+	fmt.Println(repoTable.View())
+}
+
+type ErrRepoNotFoundOrIndexed struct {
+	RepoURL string
+}
+
+func (e *ErrRepoNotFoundOrIndexed) Error() string {
+	return fmt.Sprintf("repository %s is either non-existent or has not been indexed yet", e.RepoURL)
+}
+
+func NewErrRepoNotFoundOrIndexed(repoURL string) *ErrRepoNotFoundOrIndexed {
+	return &ErrRepoNotFoundOrIndexed{
+		RepoURL: repoURL,
+	}
+}
```
Having a structured error here seems unnecessary. Typically I've only seen these to be useful when different error types need to be compared across callers of APIs. Like when using `stat` and `os`:
```go
_, err := os.Stat(path)
if os.IsNotExist(err) { 
    // handle the file at path not existin
}
```
but it doesn't seem we're doing that here. Let's remove this struct type in favor of traditional stacked error handling in Go.

jpmcb commented on cmd/insights/contributors.go:
```diff
@@ -0,0 +1,289 @@
+}
+
+func findRepositoryByOwnerAndRepoName(ctx context.Context, apiClient *client.APIClient, repoURL string) (*client.DbRepo, error) {
+	owner, repoName, err := repoquery.GetOwnerAndRepoFromURL(repoURL)
```
`GetOwnerAndRepoFromURL` should be abstracted out of the `repoquery` package.

jpmcb commented on cmd/insights/contributors.go:
```diff
@@ -0,0 +1,289 @@
+	select {
+	case <-doneChan:
+		return repoContributorInsights, nil
+	case err = <-errorChan:
+		close(errorChan)
+		return nil, err
```
Same here: only the first error gets popped off the error channel. The others get squashed.

jpmcb commented on cmd/insights/contributors.go:
```diff
@@ -0,0 +1,289 @@
+		doneChan  = make(chan bool)
+		waitGroup = new(sync.WaitGroup)
+	)
+	waitGroup.Add(4)
```
I think I'd prefer if we explicitly added `1` each time for each `go func` block. Otherwies, if another gets added, it won't be clear to other contributors what's going on here.

cecobask reviewed (commented):
Thank you for the feedback, @jpmcb!
I've updated the code based on your suggestions.
Please, review at your convenience 👍🏻

jpmcb commented:
> Thank you for the feedback, @jpmcb!

Happy to provide the feedback! This is looking good, great work. Huge lift. This is on my docket to dive back into today!

jpmcb commented on pkg/utils/arguments.go:
```diff
@@ -0,0 +1,29 @@
+)
+
+func HandleRepositoryValues(repos []string, filePath string) (map[string]bool, error) {
+	uniqueRepoURLs := make(map[string]bool)
```
I think a `map[string]struct{}` may be slightly more efficient since it essentially creates an empty member on the right side which effectively implements a "set". But that's more a suggestion: take it or leave it!

jpmcb commented on cmd/insights/contributors.go:
```diff
@@ -0,0 +1,271 @@
+	Repeat  []string
+}
+
+func (rci *repositoryContributorsInsights) RenderTable() {
```
Generally speaking, and this can be an additional issue we work on after this merges, we should also always provide a `-o / --output` flag that can be either "yaml" or "json" since we want people to be able to compose their data into other tools (like jq, grep, sed, awk, etc.)

cecobask replied:
I understand, I can refactor this to allow such an option as part of the work for https://github.com/open-sauced/pizza-cli/issues/10 👍🏻

jpmcb reviewed (commented):
This is getting very close!!!! Nice work!!!

cecobask commented:
> This is getting very close!!!! Nice work!!!

Thanks, @jpmcb! I've updated the branch with the latest changes 👍🏻 
Let me know if something else needs to be done here.

jpmcb reviewed (approved):
Great work!! Just one small nit on a regression I noticed in using a custom endpoint for the pizza service (which I typically do for testing local changes). This might be an issue with how the client generates it's endpoints. Otherwise, let's get this in! 💪🏼

cecobask commented on cmd/bake/bake.go:
Hi @jpmcb, I think it works as expected.
I haven't set up the pizza service locally but the output I'm getting indicates that the go client is using the correct URL.
This is the error output I'm getting when I use a custom URL:
```
Error: error while calling 'PizzaOvenServiceAPI.BakeARepositoryWithThePizzaOvenMicroservice' with repository "https://github.com/jpmcb/gopherlogs": Post "http://localhost:8080/v1/bake": dial tcp [::1]:8080: connect: connection refused
```
I think you're getting a 404 error because the repository might not be indexed yet, but I might be wrong.

jpmcb replied:
Good insight. I believe this is because it's attempting to use a `v1` endpoint. So, when I provide `http://localhost:8080`, it's calling into `http://localhost:8080/v1/bake` which is the actual API endpoint, not the local pizza endpoint I have setup. It's doing the right thing since the client is reaching into the routes it knows about. This is good for now unless we have a different local dev story we want to explore 👍🏼

bdougie reviewed (approved):
I have been watching this quietly in the background and can say it works as expected for me. 

re: localhost. It might be worth setting up some follow up issues to identify things we can fix and enhance after this is merged.

Something top of mind is making these contributor insights actionable with some context on  % and story telling.

Perhaps we get @isabensusan to assist in adding this to the design backlog to assist in representing this in the UI.
